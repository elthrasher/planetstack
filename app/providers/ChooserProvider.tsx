import React, { createContext, ReactNode, useState } from 'react';

export enum ChooserMode {
  NONE,
  BACKGROUND,
  ICON,
  INSTRUCTIONS,
}

export interface ChooserContextModel {
  chooserMode: ChooserMode;
  setChooserMode: (chooserMode: ChooserMode) => void;
}

export const ChooserContext = createContext<ChooserContextModel>({
  chooserMode: ChooserMode.NONE,
  setChooserMode: /* istanbul ignore next */ () => null,
});

export const ChooserProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [chooserMode, setChooserMode] = useState<ChooserMode>(ChooserMode.NONE);

  return <ChooserContext.Provider value={{ chooserMode, setChooserMode }}>{children}</ChooserContext.Provider>;
};
