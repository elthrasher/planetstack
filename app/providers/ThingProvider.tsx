import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket from 'react-use-websocket';

const socketUrl = 'wss://jdnif5uxok.execute-api.us-east-1.amazonaws.com/dev';

export interface Thing {
  id: string;
  img: string;
  x: number;
  y: number;
}

export interface ThingContextModel {
  things: Thing[];
  addThing: (thing: Thing) => void;
  removeThing: (thing: Thing) => void;
}

export const ThingContext = createContext<ThingContextModel>({
  things: [],
  addThing: () => null,
  removeThing: () => null,
});

export const ThingProvider = ({ children }: { children: ReactNode }) => {
  const [things, setThings] = useState<Thing[]>([]);
  const addThing = (thing: Thing) => setThings([...things, thing]);
  const removeThing = (thing: Thing) => setThings(things.filter((t) => t !== thing));
  const { sendMessage, sendJsonMessage, lastMessage, lastJsonMessage, readyState, getWebSocket } = useWebSocket(
    socketUrl,
    {
      onError: (e) => console.error(e),
      onOpen: () => console.log(),
      //Will attempt to reconnect on all close events, such as server shutting down
      shouldReconnect: () => true,
    },
  );
  console.log(readyState);
  return <ThingContext.Provider value={{ things, addThing, removeThing }}>{children}</ThingContext.Provider>;
};
