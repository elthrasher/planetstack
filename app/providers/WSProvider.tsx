import React, { createContext, ReactNode, useState } from 'react';
import useWebSocket, { SendMessage } from 'react-use-websocket';

import { GameStateModel } from '../../types/gamestate';
import { MessageAction } from '../../types/MessageAction';

/* istanbul ignore next */
const getSocketUrl = async () => {
  const response = await fetch('./config.json');
  const config = await response.json();
  return `${config.socketUrl}/${config.stageName}`;
};

const defaultGs: GameStateModel = { bg: 0, icons: {} };

export interface WSContextModel {
  gs: GameStateModel;
  sendMessage: SendMessage;
}

export const WSContext = createContext<WSContextModel>({
  gs: defaultGs,
  sendMessage: () => null,
});

export const WSProvider = ({ children }: { children: ReactNode }): JSX.Element => {
  const [gs, setGs] = useState(defaultGs);
  /* istanbul ignore next */
  const { sendMessage } = useWebSocket(getSocketUrl, {
    onOpen: () => sendMessage(JSON.stringify({ action: MessageAction.GET_STATE })),
    onError: (e) => console.error(e),
    onMessage: (message: MessageEvent<string>) => setGs(JSON.parse(message.data)),
    share: true,
    shouldReconnect: () => true,
  });
  return <WSContext.Provider value={{ gs, sendMessage }}>{children}</WSContext.Provider>;
};
