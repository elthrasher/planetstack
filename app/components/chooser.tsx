import React, { useContext, useState } from 'react';

import { backgrounds } from '../../types/data/backgrounds';
import { icons } from '../../types/data/icons';
import { MessageAction } from '../../types/MessageAction';
import { ChooserContext, ChooserMode } from '../providers/ChooserProvider';
import { WSContext } from '../providers/WSProvider';

export const Chooser = (): JSX.Element | null => {
  const { chooserMode, setChooserMode } = useContext(ChooserContext);
  const { sendMessage } = useContext(WSContext);

  const [bgPreview, setBGPreview] = useState(0);

  const chooseIcon = (img: number) => () => {
    setChooserMode(ChooserMode.NONE);
    sendMessage(JSON.stringify({ action: MessageAction.ADD_ICON, icon: { img, x: 400, y: 400 } }));
  };

  const chooseBG = () => {
    setChooserMode(ChooserMode.NONE);
    sendMessage(JSON.stringify({ action: MessageAction.CHANGE_BACKGROUND, bg: bgPreview }));
  };

  const goBack = () => setBGPreview(Math.max(bgPreview - 1, 0));
  const goFwd = () => setBGPreview(Math.min(bgPreview + 1, backgrounds.length - 1));
  switch (chooserMode) {
    case ChooserMode.BACKGROUND:
      return (
        <div className="container chooser">
          <i className="arrow left" data-testid="arrow-left" onClick={goBack}></i>
          <img className="background" src={backgrounds[bgPreview].path} onClick={chooseBG} />
          <i className="arrow right" data-testid="arrow-right" onClick={goFwd}></i>
        </div>
      );
    case ChooserMode.ICON:
      return (
        <div className="container chooser icons">
          {icons.map((icon, index) => (
            <img className="icon" key={index} onMouseDown={chooseIcon(index)} src={icon.path} />
          ))}
        </div>
      );
    case ChooserMode.INSTRUCTIONS:
      return (
        <div className="container chooser">
          <ul>
            <li>Select a background</li>
            <li>Click icons to add them</li>
            <li>Drag icons to your liking</li>
            <li>Right click to dismiss an icon</li>
          </ul>
        </div>
      );
    default:
      return null;
  }
};
