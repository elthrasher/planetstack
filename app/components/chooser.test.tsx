import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import { ChooserContext, ChooserMode } from '../providers/ChooserProvider';
import { Chooser } from './chooser';

const setChooser = jest.fn();

describe('Chooser component', () => {
  beforeEach(jest.resetAllMocks);

  test('No render when chooser mode is blank', () => {
    const { container } = render(
      <ChooserContext.Provider value={{ chooserMode: ChooserMode.NONE, setChooserMode: setChooser }}>
        <Chooser />
      </ChooserContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Render the background chooser', () => {
    const { container } = render(
      <ChooserContext.Provider value={{ chooserMode: ChooserMode.BACKGROUND, setChooserMode: setChooser }}>
        <Chooser />
      </ChooserContext.Provider>,
    );
    expect(container).toMatchSnapshot();
    const leftArrow = screen.getByTestId('arrow-left');
    const rightArrow = screen.getByTestId('arrow-right');
    const img = screen.getByRole('img');
    expect(img).toHaveClass('background');
    img.click();
    expect(setChooser).toHaveBeenCalledTimes(1);
    const bg1 = img.getAttribute('src');
    // navigate forward
    rightArrow.click();
    const bg2 = img.getAttribute('src');
    // navigate back
    leftArrow.click();
    const bg3 = img.getAttribute('src');
    // 1st and 2nd images are different
    expect(bg1).not.toEqual(bg2);
    // 3rd image is the same
    expect(bg1).toEqual(bg3);
  });

  test('Render the icon chooser', () => {
    const { container } = render(
      <ChooserContext.Provider value={{ chooserMode: ChooserMode.ICON, setChooserMode: setChooser }}>
        <Chooser />
      </ChooserContext.Provider>,
    );
    expect(container).toMatchSnapshot();
    const imgs = screen.getAllByRole('img');
    expect(imgs.length).toBeGreaterThan(0);
    fireEvent.mouseDown(imgs[0]);
    expect(setChooser).toHaveBeenCalledTimes(1);
  });

  test('Render instructions', () => {
    const { container } = render(
      <ChooserContext.Provider value={{ chooserMode: ChooserMode.INSTRUCTIONS, setChooserMode: setChooser }}>
        <Chooser />
      </ChooserContext.Provider>,
    );
    expect(container).toHaveTextContent('Select a background');
    expect(container).toMatchSnapshot();
  });
});
