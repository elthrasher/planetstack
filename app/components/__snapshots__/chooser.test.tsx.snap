// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chooser component No render when chooser mode is blank 1`] = `<div />`;

exports[`Chooser component Render instructions 1`] = `
<div>
  <div
    class="container chooser"
  >
    <ul>
      <li>
        Select a background
      </li>
      <li>
        Click icons to add them
      </li>
      <li>
        Drag icons to your liking
      </li>
      <li>
        Right click to dismiss an icon
      </li>
    </ul>
  </div>
</div>
`;

exports[`Chooser component Render the background chooser 1`] = `
<div>
  <div
    class="container chooser"
  >
    <i
      class="arrow left"
      data-testid="arrow-left"
    />
    <img
      class="background"
      src="assets/Alpine Planet 1280x720/1280x720 Alpine - Dry.png"
    />
    <i
      class="arrow right"
      data-testid="arrow-right"
    />
  </div>
</div>
`;

exports[`Chooser component Render the icon chooser 1`] = `
<div>
  <div
    class="container chooser icons"
  >
    <img
      class="icon"
      src="assets/RandomBuildings/B01.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B02.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B03.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B04.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B05.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B06.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B07.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B08.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B09.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B10.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B11.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B12.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B13.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B14.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B15.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B16.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B17.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B18.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B19.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B20.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B21.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B22.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B23.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B24.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B25.png"
    />
    <img
      class="icon"
      src="assets/RandomBuildings/B26.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret01.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret02.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret03.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret04.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret05.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret06.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret07.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret08.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret09.png"
    />
    <img
      class="icon"
      src="assets/DifferentTurrets/Turret10.png"
    />
    <img
      class="icon"
      src="assets/Bullets/EnemyB01.png"
    />
    <img
      class="icon"
      src="assets/Bullets/EnemyB02.png"
    />
    <img
      class="icon"
      src="assets/Bullets/P02.png"
    />
    <img
      class="icon"
      src="assets/Bullets/P03.png"
    />
    <img
      class="icon"
      src="assets/Bullets/P04.png"
    />
    <img
      class="icon"
      src="assets/Bullets/P05.png"
    />
    <img
      class="icon"
      src="assets/MiniAsteroids/01.png"
    />
    <img
      class="icon"
      src="assets/MiniAsteroids/02.png"
    />
    <img
      class="icon"
      src="assets/MiniAsteroids/03.png"
    />
  </div>
</div>
`;
